File: Data Structures & Algorithms/clone-graph/submission-11.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> neighbors;
    Node() : val(0), neighbors(vector<Node*>()) {}
    Node(int _val) : val(_val), neighbors(vector<Node*>()) {}
    Node(int _val, vector<Node*> _neighbors) : val(_val), neighbors(_neighbors) {}
};
*/

class Solution {
public:
    Node* cloneGraph(Node* node) {
        // empty graph
        if (node == nullptr) return nullptr;

        // mp[original] = cloned copy of original
        unordered_map<Node*, Node*> mp;

        // bfs queue over original nodes
        queue<Node*> q;

        // create the clone of the starting node (same value, empty neighbors for now)
        mp[node] = new Node(node->val);

        // start bfs from the input node
        q.push(node);

        while (!q.empty()) {
            // take one original node to process
            Node* cur = q.front();
            q.pop();

            // loop over all neighbors of this original node
            // (this runs as many times as cur->neighbors.size())
            for (Node* nei : cur->neighbors) {

                // if we have not cloned this neighbor yet, clone it once
                if (mp.find(nei) == mp.end()) {
                    mp[nei] = new Node(nei->val); // clone neighbor node value
                    q.push(nei);                  // later we will process its neighbors too
                }

                // connect the cloned current node to the cloned neighbor
                mp[cur]->neighbors.push_back(mp[nei]);
            }
        }

        // return the cloned version of the start node
        return mp[node];
    }
};

/*
time complexity: O(v + e)
- v nodes, e edges, each processed once during bfs

space complexity: O(v)
- hashmap + queue in worst case
*/

File: Data Structures & Algorithms/valid-tree/submission-11.py
class Solution:
    def validTree(self, n: int, edges: List[List[int]]) -> bool:
        if len(edges) != n-1:
            return False
        

        graph = defaultdict(list)
        for u,v in edges:
            graph[u].append(v)
            graph[v].append(u)
        
        visited = {0}
        queue = deque([(0,-1)])

        while queue: # We process level-by-level - all neighbors of node 0, 
        # then all neighbors of node 1, etc.
            node,parent = queue.popleft()
            for neighbor in graph[node]:
                if neighbor == parent:
                    continue
                if neighbor in visited: # when we encounter a visited node 
                # that's NOT the parent, it's a cycle!
                    return False

                visited.add(neighbor)
                queue.append((neighbor,node))
        
        return len(visited) == n







        